pkg/registry/core/service/storage/alloc.go

```diff
@@ -414,13 +415,17 @@ func (al *Allocators) allocIPs(service *api.Service, toAlloc map[api.IPFamily]st
 				allocatedIP, err = allocator.AllocateNext()
 			}
 			if err != nil {
-				return allocated, errors.NewInternalError(fmt.Errorf("failed to allocate a serviceIP: %v", err))
+				if !errors.Is(err, ipallocator.ErrFull) {
+					el := field.ErrorList{field.Invalid(field.NewPath("spec", "clusterIPs"), service.Spec.ClusterIPs, fmt.Sprintf("failed to allocate IP: %v", err))}
+					return allocated, apierrors.NewInvalid(api.Kind("Service"), service.Name, el)
```

Comment:
If the invalid field is empty name or generateName with unexpected char, the user can not get correct error message?